ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/HARouting.java

```diff
@@ -282,45 +281,48 @@ static RoutingResult executeOrRouteQuery(
             routingOptions.getIsDebugRequest() ? node : null));
-    if (node.isLocal()) {
-      try {
-        LOG.debug("Query {} executed locally at host {} at timestamp {}.",
-                  statement.getStatementText(), node.location(), System.currentTimeMillis());
-        pullQueryMetrics
+
+    final LinkedList<KsqlPartitionLocation> remainingPartitions = new LinkedList<>();
+    for (int i = 0; i < locations.size(); i++) {
+      if (node.isLocal()) {
+        try {
+          LOG.debug("Query {} executed locally at host {} at timestamp {}.",
+              statement.getStatementText(), node.location(), System.currentTimeMillis());
+          pullQueryMetrics
             .ifPresent(queryExecutorMetrics -> queryExecutorMetrics.recordLocalRequests(1));
-        pullPhysicalPlan.execute(locations, pullQueryQueue,  rowFactory);
-        return RoutingResult.SUCCESS;
-      } catch (StandbyFallbackException e) {
-        LOG.warn("Error executing query locally at node {}. Falling back to standby state which "
-                + "may return stale results", node, e.getCause());
-        return RoutingResult.STANDBY_FALLBACK;
-      } catch (Exception e) {
-        throw new KsqlException(
+          pullPhysicalPlan.execute(ImmutableList.of(locations.get(i)), pullQueryQueue, rowFactory);
+        } catch (StandbyFallbackException e) {
+          LOG.warn("Error executing query locally at node {}. Falling back to standby state which "
+              + "may return stale results", node, e.getCause());
+          remainingPartitions.add(locations.get(i));
+        } catch (Exception e) {
+          throw new KsqlException(
             String.format("Error executing query locally at node %s: %s", node.location(),
-                e.getMessage()),
+              e.getMessage()),
             e
-        );
-      }
-    } else {
-      try {
-        LOG.debug("Query {} routed to host {} at timestamp {}.",
-            statement.getStatementText(), node.location(), System.currentTimeMillis());
-        pullQueryMetrics
+          );
+        }
+      } else {
+        try {
+          LOG.debug("Query {} routed to host {} at timestamp {}.",
+              statement.getStatementText(), node.location(), System.currentTimeMillis());
+          pullQueryMetrics
             .ifPresent(queryExecutorMetrics -> queryExecutorMetrics.recordRemoteRequests(1));
-        forwardTo(node, locations, statement, serviceContext, pullQueryQueue, rowFactory,
-            outputSchema);
-        return RoutingResult.SUCCESS;
-      } catch (StandbyFallbackException e) {
-        LOG.warn("Error forwarding query to node {}. Falling back to standby state which may "
-                + "return stale results", node.location(), e.getCause());
-        return RoutingResult.STANDBY_FALLBACK;
-      } catch (Exception e) {
-        throw new KsqlException(
+          forwardTo(node, ImmutableList.of(locations.get(i)), statement, serviceContext,
```

Comment:
if I'm reading this correctly, we're no longer batching requests so this might multiply the amount of requests to other nodes by the number of remaining partitions - right?

Reply:
Yea, good point and that's a concern, that's also why the PR was marked "not ready" :-).

Ideally we would want the add support to recover individual partitions (as opposed to all partitions of a node) fast, but also keep the request efficiency we have current (where requests for multiple nodes are batched per node). I can see 3 basic ways of implementing something close to this, as discussed below. The first approach is related to the PR update of today.

1) Don't batch but parallelize partition fetching (this update)
- All partition fetching funnels through the executorservice. Means that even though they are separate requests, they are basically back-to-back on the wire on the same connection per node. 
- The advantage of this mode is, it's simple, and it can be extended quite comfortably to a mode where we re-try failed partitions immediately without having to wait for other partitions (from the same node) to be fetched, something we currently do because we have one single batched request per node. 

2) Batch requests per node and return IDs of failed partitions
- Requires changing the RPC interface we currently have in SimpleKsqlClient
- Head-of-line blocking problem: we can't retry failed partitions before the entire batch completed

3) Batch requests per node but notify coordinator about failures asynchronously
- Efficient: We can retry failed partitions immediately without having to wait for the batch to complete
- Complex logic, needs async notification channel, and we need to keep state per partition

Reply:
> All partition fetching funnels through the executorservice. Means that even though they are separate requests, they are basically back-to-back on the wire on the same connection per node.

Technically, if we used http2, they would be multiplexed on the same connection, which could save us some of that efficiency.

> Head-of-line blocking problem: we can't retry failed partitions before the entire batch completed

We always wait until the entire round of requests are done to fallback, which isn't super efficient, but it's simple.  So this wouldn't be any worse, but you could say it prevents us from being more sophisticated in the future.

One thing that also occurs to me is that a partition is not the finest grained unit of data.  Even for a single partition, we could fail mid way after having returned some rows and decide we want to start all over again.  Just pointing this out that we'll never be able (without much more work) to avoid repeats of data during failures.  Obviously, the finer grained the better, but also, especially with heartbeating, failures happen pretty rarely.

If we decided that we wanted to go this route, the least complex solution would be to avoid any kind of signaling about what partition succeeded and what didn't (since a whole request could always fail) and just have one request per partition over http2.

Reply:
### Performance concern

Thanks, all. I agree that firing off a request per partition likely has more overhead than batching all the partitions into one request per node. Also that switching to HTTP/2 in the future would diminish this overhead.

I think we'd like to know about the potential performance impact before merging this PR. If it's bad, we should have a plan to immediately switch to HTTP/2 or re-design this feature to multiplex the partition responses. And of course, we would also want to re-evaluate the performance with those fixes in place _also_ to make sure they actually fix the problem.

The punchline is, this is a legitimate performance concern, and we should perform a benchmark test to validate it.

### Failure/retry

Alan is correct. Any time we retry part of the query internally, we would repeat part of the response. This is also true today. There's not much we can do about it in the routing layer. The only way to avoid it is to never retry internally or to buffer the entire query result in the routing node before returning anything to the user.

Reply:
Separate feedback: we shouldn't be blocking right away on `future.get`. Instead we should just repeatedly check async whether the futures have succeeded or failed and retry the failed ones right away.

Reply:
A background question: Under what scenarios is it possible for a partition on a node to fail but other partitions on that same node don't fail? What kind of failures are these? And in these cases, do we want to route the query to a different node that hosts the failed partition or do we want to retry on that same node? If the node is not down and the failure is an ephemeral fluke, it is cheaper to retry on the same node instead of rerouting.